### ISSUE TITLE
Fitting with generators and sample weights crashes when batch size varies between steps with 3D output

### ISSUE THREAD

**Issue by Zahlii** [tensorflow/tensorflow#57158]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.9.1

### Custom Code

No

### OS Platform and Distribution

MacOS 12.5

### Python version

3.7.3

### CUDA/cuDNN version

-

### GPU model and memory

-

### Current Behaviour?

```shell
When training a model with a generator (from dataset.from_tensor_slices) and including sample weights where the batch size varies, an error is thrown.

Note: This test for us previously worked as expected in TF<2.6!
```

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf

x = np.random.random((200, 20, 20, 3)) * 10
y = x.dot(np.random.random((3, 3)))
x = x.astype(np.uint8)

model = tf.keras.Sequential()
model.add(
    tf.keras.layers.Conv2D(
        filters=3,
        kernel_size=3,
        activation="relu",
        input_shape=(20, 20, 3),
        padding="same",
    )
)
model.summary()
model.compile(optimizer="adam", loss="mse")

# Works
gen = tf.data.Dataset.from_tensor_slices((x, y)).batch(16).repeat()
model.fit(gen, epochs=2, steps_per_epoch=10)

# Works
gen3 = tf.data.Dataset.from_tensor_slices((x, y, np.ones((200,)))).batch(20).repeat()
model.fit(gen3, epochs=2, steps_per_epoch=10)

# Doesnt work
gen2 = tf.data.Dataset.from_tensor_slices((x, y, np.ones((200,)))).batch(16).repeat()
model.fit(gen2, epochs=2, steps_per_epoch=10)
```

### Relevant log output

```shell
grads = tape.gradient(loss, var_list, grad_loss)
Node: 'gradient_tape/mean_squared_error/weighted_loss/BroadcastGradientArgs'
Incompatible shapes: [16,20,20] vs. [16]
	 [[{{node gradient_tape/mean_squared_error/weighted_loss/BroadcastGradientArgs}}]] [Op:__inference_train_function_682]
```
</details>

**Comment by Zahlii** [comment]
Interestingly, the following works (i.e. "normal" output instead of 3D)

    import numpy as np
    import tensorflow as tf

    x = np.random.random((200, 20, 20, 3)) * 10
    y = x.dot(np.random.random((3, 3))).sum(axis=(1,2,3))
    x = x.astype(np.uint8)
    print(x.shape, y.shape)
    model = tf.keras.Sequential()
    model.add(
    tf.keras.layers.Conv2D(
        filters=3,
        kernel_size=3,
        activation="relu",
        input_shape=(20, 20, 3),
        padding="same",
    )
    )
    model.add(tf.keras.layers.Flatten())
    model.add(tf.keras.layers.Dense(1))
    model.summary()
    model.compile(optimizer="adam", loss="mse")

    # Works
    gen = tf.data.Dataset.from_tensor_slices((x, y)).batch(16).repeat()
    model.fit(gen, epochs=2, steps_per_epoch=10)

    # Works
    gen3 = tf.data.Dataset.from_tensor_slices((x, y, np.ones((200,)))).batch(20).repeat()
    model.fit(gen3, epochs=2, steps_per_epoch=10)

    # Works
    gen2 = tf.data.Dataset.from_tensor_slices((x, y, np.ones((200,)))).batch(16).repeat()
    model.fit(gen2, epochs=2, steps_per_epoch=10)

**Comment by sushreebarsa** [comment]
@Zahlii I tried to replicate the issue on colab, please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/f05f31c10c6155387aedd5951091979b/57158.ipynb#scrollTo=gkT5Zc0M22pg)and confirm the same. Thank you!
